File: exampleadapters/src/main/java/foo/bar/example/asafadapters/feature/playlist/PlaylistAdvancedModel.java
package foo.bar.example.asafadapters.feature.playlist;


import co.early.asaf.core.Affirm;
import co.early.asaf.core.WorkMode;
import co.early.asaf.core.logging.Logger;
import co.early.asaf.core.observer.ObservableImp;
import co.early.asaf.adapters.SimpleChangeAwareList;
import co.early.asaf.adapters.UpdateSpec;
import co.early.asaf.adapters.Updateable;

import static foo.bar.example.asafadapters.feature.playlist.RandomTrackGeneratorUtil.generateRandomColourResource;

/**
 *
 */
public class PlaylistAdvancedModel extends ObservableImp implements Updateable {

    private static String TAG = PlaylistAdvancedModel.class.getSimpleName();

    private final Logger logger;
    private SimpleChangeAwareList<Track> trackList = new SimpleChangeAwareList<>();


    public PlaylistAdvancedModel(WorkMode workMode, Logger logger) {
        super(workMode, logger);
        this.logger = Affirm.notNull(logger);
    }

    public void addNewTrack() {
        logger.i(TAG, "addNewTrack()");
        trackList.add(new Track(generateRandomColourResource()));
        notifyObservers();
    }

    public void removeTrack(int index) {
        logger.i(TAG, "removeTrack() " + index);
        checkIndex(index);
        trackList.remove(index);
        notifyObservers();
    }

    public void removeAllTracks() {
        logger.i(TAG, "removeAllTracks()");
        trackList.clear();
        notifyObservers();
    }

    public void increasePlaysForTrack(int index) {
        logger.i(TAG, "increasePlaysForTrack() " + index);
        getTrack(index).increasePlaysRequested();
        trackList.makeAwareOfDataChange(index);
        notifyObservers();
    }

    public void decreasePlaysForTrack(int index) {
        logger.i(TAG, "decreasePlaysForTrack() " + index);
        getTrack(index).decreasePlaysRequested();
        trackList.makeAwareOfDataChange(index);
        notifyObservers();
    }

    public Track getTrack(int index) {
        checkIndex(index);
        return trackList.get(index);
    }

    public int getTrackListSize() {
        return trackList.size();
    }

    private void checkIndex(int index) {
        if (trackList.size() == 0) {
            throw new IndexOutOfBoundsException("tracklist has no items in it, can not get index:" + index);
        } else if (index < 0 || index > trackList.size() - 1) {
            throw new IndexOutOfBoundsException("tracklist index needs to be between 0 and " + (trackList.size() - 1) + " not:" + index);
        }
    }

    @Override
    public UpdateSpec getAndClearMostRecentUpdateSpec() {
        return trackList.getAndClearMostRecentUpdateSpec();
    }
}

File: example-kt-03adapters/src/main/java/foo/bar/example/foreadapterskt/feature/playlist/PlaylistAdvancedModel.kt
package foo.bar.example.foreadapterskt.feature.playlist


import co.early.fore.adapters.ChangeAwareArrayList
import co.early.fore.adapters.ChangeAwareList
import co.early.fore.adapters.UpdateSpec
import co.early.fore.adapters.Updateable
import co.early.fore.core.WorkMode
import co.early.fore.kt.core.logging.Logger
import co.early.fore.core.observer.Observable
import co.early.fore.core.time.SystemTimeWrapper
import co.early.fore.kt.core.observer.ObservableImp
import foo.bar.example.foreadapterskt.feature.playlist.RandomTrackGeneratorUtil.generateRandomColourResource
import java.util.ArrayList

/**
 * Copyright © 2019 early.co. All rights reserved.
 */
class PlaylistAdvancedModel(
        systemTimeWrapper: SystemTimeWrapper,
        workMode: WorkMode,
        private val logger: Logger
) : Observable by ObservableImp(workMode, logger),
    Updateable {

    private val trackList: ChangeAwareList<Track>

    val trackListSize: Int
        get() = trackList.size


    init {
        trackList = ChangeAwareArrayList(systemTimeWrapper)
    }

    fun addNewTrack() {
        logger.i("addNewTrack()")
        trackList.add(Track(generateRandomColourResource()))
        notifyObservers()
    }

    fun removeTrack(index: Int) {
        logger.i("removeTrack() $index")
        checkIndex(index)
        trackList.removeAt(index)
        notifyObservers()
    }

    fun removeAllTracks() {
        logger.i("removeAllTracks()")
        trackList.clear()
        notifyObservers()
    }

    fun increasePlaysForTrack(index: Int) {
        logger.i("increasePlaysForTrack() $index")
        getTrack(index).increasePlaysRequested()
        trackList.makeAwareOfDataChange(index)
        notifyObservers()
    }

    fun decreasePlaysForTrack(index: Int) {
        logger.i("decreasePlaysForTrack() $index")
        getTrack(index).decreasePlaysRequested()
        trackList.makeAwareOfDataChange(index)
        notifyObservers()
    }

    fun getTrack(index: Int): Track {
        checkIndex(index)
        return trackList[index]
    }

    fun add5NewTracks() {
        logger.i("add5NewTracks()")
        val newTracks = ArrayList<Track>()
        for (ii in 0..4) {
            newTracks.add(Track(generateRandomColourResource()))
        }
        trackList.addAll(0, newTracks)
        notifyObservers()
    }

    fun remove5Tracks() {
        logger.i("remove5Tracks()")
        if (trackListSize > 4) {
            trackList.removeRange(0, 5)
            notifyObservers()
        }
    }

    private fun checkIndex(index: Int) {
        if (trackList.size == 0) {
            throw IndexOutOfBoundsException("tracklist has no items in it, can not get index:$index")
        } else if (index < 0 || index > trackList.size - 1) {
            throw IndexOutOfBoundsException("tracklist index needs to be between 0 and " + (trackList.size - 1) + " not:" + index)
        }
    }

    override fun getAndClearLatestUpdateSpec(maxAgeMs: Long): UpdateSpec {
        return trackList.getAndClearLatestUpdateSpec(maxAgeMs)
    }

}
